src/components/utils/src/pipe_qt.cc

```diff
@@ -0,0 +1,174 @@
+/*
+ * Copyright (c) 2015, Ford Motor Company
+ * All rights reserved.
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions are met:
+ *
+ * Redistributions of source code must retain the above copyright notice, this
+ * list of conditions and the following disclaimer.
+ *
+ * Redistributions in binary form must reproduce the above copyright notice,
+ * this list of conditions and the following
+ * disclaimer in the documentation and/or other materials provided with the
+ * distribution.
+ *
+ * Neither the name of the Ford Motor Company nor the names of its contributors
+ * may be used to endorse or promote products derived from this software
+ * without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+ * LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+ * CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+ * SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ * INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ * CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ * ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ * POSSIBILITY OF SUCH DAMAGE.
+ */
+#include <string>
+#include <algorithm>
+
+#include "utils/winhdr.h"
+#include <QtNetwork>
+#include "utils/pipe.h"
+
+namespace std {
+
+template <>
+void swap<utils::Pipe>(utils::Pipe& lhs, utils::Pipe& rhs) {
+  lhs.Swap(rhs);
+}
+
+}  // namespace std
+
+namespace utils {
+
+class Pipe::Impl {
+ public:
+  Impl();
+  explicit Impl(QLocalSocket pipe);
+  ~Impl();
+
+  Impl(Impl& rh);
+  Impl& operator=(Impl& rh);
+
+  bool Valid() const;
+
+  bool Create(const std::string& name);
+
+  bool Open();
+  bool Close();
+
+  ssize_t Write(const char* buf, size_t length);
+
+ private:
+  QLocalSocket* pipe_;
+  QString name_;
+};
+
+}  // namespace utils
+
+utils::Pipe::Pipe() : impl_(new Pipe::Impl()) {}
+
+utils::Pipe::~Pipe() { delete impl_; }
+
+utils::Pipe::Pipe(Pipe& rh) {
+  impl_ = new Pipe::Impl();
+  *impl_ = *rh.impl_;
+}
+
+utils::Pipe& utils::Pipe::operator=(Pipe& rh) {
```

Comment:
const cannot be used here since this operator will actually change the `rh`. This Pipe class works as auto_ptr. On assignment it will move the ownership of the impl.